File: ds_algos_primer/javascript/StacksAndQueues.js
/**
 *   Title: StacksAndQueues.js
 *
 *   This file contains the template for the Stack & Queue exercises in the
 *   DS & Algos Primer. Fill in the exercises here and refer to
 *   StackAndQueueSolutions.js for the complete code samples.
 *
 *   Execution: node StacksAndQueues.js
 */

/**
 * Exercise 1.1: Implement a Stack of integers using a Linked List
 */
function MyStack() {

    /**
     * A simple singly-linked node class
     *
     * @param{number} n
     */
    function Node(n) {
        this.val = n;
        this.next = null;
    }

    /**
     * Add an item to the top of the stack
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @param{number} val
     */
    this.push = function(val) {
        // INSERT YOUR CODE HERE
    }

    /**
     * Remove an item from the top of the stack
     *
     * Time Complexity:
     * Space Complextiy:
     *
     * @return{number}
     */
    this.pop = function() {
        // INSERT YOUR CODE HERE
    }

    /*
     * Get the size of the stack
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @return{number}
     */
    this.size = function() {
        // INSERT YOUR CODE HERE
    }
}

/*
 * Exercise 1.2: Implement a Queue of integers using a Linked List
 */
function MyQueue() {

    /**
     * Simple Doubly Linked List node class
     *
     * @param{number} n
     */
    function Node(n) {
        this.val = n;
        this.next = null;
        this.prev = null;
    }

    /**
     * Add an item to the beginning of the queue
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @param{number} val
     */
    this.enqueue = function(val) {
        // INSERT YOUR CODE HERE
    }

    /**
     * Remove an item from the end of the queue
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @return{number}
     */
    this.dequeue = function() {
        // INSERT YOUR CODE HERE
    }

    /**
     * Get the size of the stack
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @return{number}
     */
    this.size = function() {
        // INSERT YOUR CODE HERE
    }
}

/**
 * Exercise 1.3: Given a stack, find the nth element in the stack
 *
 * Time Complexity:
 * Space Complexity:
 *
 * @param{number[]} stack
 * @param{number} n
 * @return{number}
 */
var nthElementInStack = function(stack, n) {
    // INSERT YOUR CODE HERE
}

/**
 * Exercise 1.4: Implement a Stack using Queues
 */

function StackFromQueues() {

    /**
     * Push an item onto the stack
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @param{number} x
     */
    this.push = function(x) {
        // INSERT YOUR CODE HERE
    }

    /**
     * Pop an item from the stack
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @return{number}
     */
    this.pop = function() {
        // INSERT YOUR CODE HERE
    }

    /**
     * Return the top item on the stack
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @return{number}
     */
    this.top = function() {
        // INSERT YOUR CODE HERE
    }

    /**
     * Is the stack empty
     *
     * Time Complexity:
     * Space Complexity:
     *
     * @return{boolean}
     */
    this.empty = function() {
        // INSERT YOUR CODE HERE
    }
}

/**
 * Exercise 2.1: Valid Parentheses
 *
 * Time Complexity:
 * Space Complexity:
 *
 * @param{string} s
 * @return{boolean}
 */
var validParentheses = function(s) {
    // INSERT YOUR CODE HERE
}

/**
 * Exercise 2.2: Basic Calculator
 *
 * Time Complexity:
 * Space Complexity:
 *
 * @param{string} s
 * @return{number}
 */
var basicCalculator = function(s) {
    // INSERT YOUR CODE HERE
}

// Test cases
var tester = function() {
    // INSERT YOUR TEST CASES HERE
}

tester();

File: ds_algos_primer/python/stacks_and_queues.py
"""
Title: Stack & Queue Solutions

This file contains the template for the Stack & Queue exercises in the
DS & Algos Primer. Fill in the exercises here and refer to
stack_and_queue_solutions.py for the complete code samples.

Execution: python stacks_and_queues.py
"""

import queue
from typing import List

"""
Exercise 1.1: Implement a Stack of integers using a Linked List
"""
class MyStack:

    """
    A simple singly-linked node class
    """
    class _Node:
        def __init__(self, n: int):
            self.val = n
            self.next = None

    """
    Constructor
    """
    def __init__(self):
        # INSERT YOUR SOLUTION HERE

    """
    Add an item to the top of the stack

    Time Complexity:
    Space Complexity:
    """
    def push(self, val: int):
        # INSERT YOUR SOLUTION HERE

    """
    Remove an item from the top of the stack

    Time Complexity:
    Space Complextiy:
    """
    def pop(self) -> int:
        # INSERT YOUR SOLUTION HERE

    """
    Get the size of the stack

    Time Complexity:
    Space Complexity:
    """
    def size(self) -> int:
        # INSERT YOUR SOLUTION HERE

"""
Exercise 1.2: Implement a Queue of integers using a Linked List
"""
class MyQueue:

    """
    A simple doubly-linked node class
    """
    class _Node:
        def __init__(self, n: int):
            self.val = n
            self.next = None
            self.prev = None

    """
    Constructor
    """
    def __init__(self):
        # INSERT YOUR SOLUTION HERE

    """
    Add an item to the beginning of the queue

    Time Complexity:
    Space Complexity:
    """
    def enqueue(self, val: int):
        # INSERT YOUR SOLUTION HERE

    """
    Remove an item from the end of the queue

    Time Complexity:
    Space Complexity:
    """
    def dequeue(self) -> int:
        # INSERT YOUR SOLUTION HERE

    """
    Get the size of the stack

    Time Complexity:
    Space Complexity:
    """
    def size(self) -> int:
        # INSERT YOUR SOLUTION HERE

"""
Exercise 1.3: Given a stack, find the nth element in the stack

Time Complexity:
Space Complexity:
"""
def nth_element_in_stack(stack: List[int], n: int) -> int:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 1.4: Implement a Stack using Queues
"""
class StackFromQueues:

    """
    Constructor
    """
    def __init__(self):
        # INSERT YOUR SOLUTION HERE

    """
    Push an item onto the stack

    Time Complexity:
    Space Complexity:
    """
    def push(self, x: int) -> None:
        # INSERT YOUR SOLUTION HERE

    """
    Pop an item from the stack

    Time Complexity:
    Space Complexity:
    """
    def pop(self) -> int:
        # INSERT YOUR SOLUTION HERE

    """
    Return the top item on the stack

    Time Complexity:
    Space Complexity:
    """
    def top(self) -> int:
        # INSERT YOUR SOLUTION HERE

    def empty(self) -> bool:
        # INSERT YOUR SOLUTION HERE

"""
Exercise 2.1: Valid Parentheses

Time Complexity:
Space Complexity:
"""
def valid_parentheses(s: str) -> bool:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 2.2: Basic Calculator

Time Complexity:
Space Complexity:
"""
def basic_calculator(s: str) -> int:
    # INSERT YOUR SOLUTION HERE

if __name__ == '__main__':
    # INSERT YOUR TEST CASES HERE
